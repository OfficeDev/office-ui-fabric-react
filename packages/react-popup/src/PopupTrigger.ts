export * from './components/PopupTrigger/index';
