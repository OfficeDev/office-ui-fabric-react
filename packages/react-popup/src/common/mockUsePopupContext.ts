import { PopupContextValue, usePopupContext } from '../popupContext';

/**
 * A test utility to mock the usePopupContext hook that relies on context selector
 * Don't forget to call `jest.mock(**\/popupContext.ts)` while using this
 * @param options Popup context values to set for testing
 */
export const mockPopupContext = (options: Partial<PopupContextValue> = {}) => {
  const mockContext: PopupContextValue = {
    target: null,
    triggerRef: { current: null },
    contentRef: { current: null },
    open: false,
    setOpen: jest.fn(),
    mountNode: null,
    openOnContext: false,
    openOnHover: false,
    ...options,
  };
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (usePopupContext as jest.Mock).mockImplementation((selector: (context: PopupContextValue) => any) => {
    return selector(mockContext);
  });
};
