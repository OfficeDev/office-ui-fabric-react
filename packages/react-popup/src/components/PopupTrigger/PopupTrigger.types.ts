import * as React from 'react';

/**
 * PopupTrigger Props
 */
export interface PopupTriggerProps {
  children: React.ReactElement;
}

/**
 * PopupTrigger State
 */
export interface PopupTriggerState extends PopupTriggerProps {}
