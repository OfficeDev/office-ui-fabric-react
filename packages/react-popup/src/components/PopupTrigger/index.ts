export * from './PopupTrigger';
export * from './PopupTrigger.types';
export * from './renderPopupTrigger';
export * from './usePopupTrigger';
