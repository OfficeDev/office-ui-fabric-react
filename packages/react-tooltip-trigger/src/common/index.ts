export * from './TooltipContext.types';
export * from './TooltipManager.types';
export * from './TooltipProps.types';
export * from './useTooltipContext';
