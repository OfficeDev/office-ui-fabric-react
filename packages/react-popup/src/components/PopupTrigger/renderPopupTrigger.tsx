import { PopupTriggerState } from './PopupTrigger.types';

/**
 * Render the final JSX of PopupTrigger
 */
export const renderPopupTrigger = (state: PopupTriggerState): JSX.Element => {
  return state.children;
};
