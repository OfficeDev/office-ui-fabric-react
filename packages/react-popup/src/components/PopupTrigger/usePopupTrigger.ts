import * as React from 'react';
import {
  makeMergeProps,
  useMergedRefs,
  useEventCallback,
  shouldPreventDefaultOnKeyDown,
} from '@fluentui/react-utilities';
import { PopupTriggerProps, PopupTriggerState } from './PopupTrigger.types';
import { usePopupContext } from '../../popupContext';

const mergeProps = makeMergeProps<PopupTriggerState>({});

/**
 * Create the state required to render PopupTrigger.
 *
 * The returned state can be modified with hooks such as usePopupTriggerStyles,
 * before being passed to renderPopupTrigger.
 *
 * @param props - props from this instance of PopupTrigger
 * @param defaultProps - (optional) default prop values provided by the implementing type
 */
export const usePopupTrigger = (props: PopupTriggerProps, defaultProps?: PopupTriggerProps): PopupTriggerState => {
  const setOpen = usePopupContext(context => context.setOpen);
  const open = usePopupContext(context => context.open);
  const triggerRef = usePopupContext(context => context.triggerRef);
  const openOnHover = usePopupContext(context => context.openOnHover);
  const openOnContext = usePopupContext(context => context.openOnContext);

  const state = mergeProps(
    {
      children: (null as unknown) as React.ReactElement,
    },
    defaultProps,
    props,
  );

  const onContextMenu = useEventCallback((e: React.MouseEvent<HTMLElement>) => {
    if (openOnContext) {
      e.preventDefault();
      setOpen(e, true);
    }
    child.props?.onContextMenu?.(e);
  });

  const onClick = useEventCallback((e: React.MouseEvent<HTMLElement>) => {
    if (!openOnContext) {
      setOpen(e, !open);
    }
    child.props?.onClick?.(e);
  });

  const onKeyDown = useEventCallback((e: React.KeyboardEvent<HTMLElement>) => {
    if (shouldPreventDefaultOnKeyDown(e)) {
      e.preventDefault();
      (e.target as HTMLElement)?.click();
    }

    child.props?.onKeyDown?.(e);
  });

  const onMouseEnter = useEventCallback((e: React.MouseEvent<HTMLElement>) => {
    if (openOnHover && !openOnContext) {
      setOpen(e, true);
    }
    child.props?.onMouseEnter?.(e);
  });

  const onMouseLeave = useEventCallback((e: React.MouseEvent<HTMLElement>) => {
    if (openOnHover && !openOnContext) {
      setOpen(e, false);
    }
    child.props?.onMouseEnter?.(e);
  });

  const child = React.Children.only(state.children);
  state.children = React.cloneElement(child, {
    'aria-haspopup': 'true',
    ...child.props,

    // These event handlers must handle original props
    onClick,
    onMouseEnter,
    onKeyDown,
    onMouseLeave,
    onContextMenu,
    ref: useMergedRefs(child.props.ref, triggerRef),
  });

  return state;
};
