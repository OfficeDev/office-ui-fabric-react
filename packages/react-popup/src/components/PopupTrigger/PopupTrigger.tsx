import * as React from 'react';
import { usePopupTrigger } from './usePopupTrigger';
import { PopupTriggerProps } from './PopupTrigger.types';
import { renderPopupTrigger } from './renderPopupTrigger';

/**
 * PopupTrigger component
 */
export const PopupTrigger: React.FC<PopupTriggerProps> = props => {
  const state = usePopupTrigger(props);

  return renderPopupTrigger(state);
};

PopupTrigger.displayName = 'PopupTrigger';
