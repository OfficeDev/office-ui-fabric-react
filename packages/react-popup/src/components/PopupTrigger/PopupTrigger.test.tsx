import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { PopupTrigger } from './PopupTrigger';
import * as renderer from 'react-test-renderer';
import { mockPopupContext } from '../../common/mockUsePopupContext';

jest.mock('../../popupContext');

describe('PopupTrigger', () => {
  beforeEach(() => {
    mockPopupContext();
  });

  /**
   * Note: see more visual regression tests for PopupTrigger in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(
      <PopupTrigger>
        <button>Popup trigger</button>
      </PopupTrigger>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.each([
    ['onClick', fireEvent.click],
    ['onContextMenu', fireEvent.contextMenu],
    ['onKeyDown', fireEvent.keyDown],
    ['onMouseEnter', fireEvent.mouseEnter],
    ['onMouseLeave', fireEvent.mouseLeave],
  ])('should keep the original %s handler', (handler, triggerEvent) => {
    // Arrange
    const spy = jest.fn();
    const { getByRole } = render(
      <PopupTrigger>
        <button {...{ [handler]: spy }}>Trigger</button>
      </PopupTrigger>,
    );

    // Act
    triggerEvent(getByRole('button'));

    // Assert
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should set aria-haspopup on trigger element', () => {
    // Arrange
    const { getByRole } = render(
      <PopupTrigger>
        <button>Trigger</button>
      </PopupTrigger>,
    );

    // Assert
    expect(getByRole('button').getAttribute('aria-haspopup')).toEqual('true');
  });
});
