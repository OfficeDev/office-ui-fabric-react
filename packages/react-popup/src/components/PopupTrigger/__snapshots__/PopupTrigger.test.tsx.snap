// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupTrigger renders a default state 1`] = `
<button
  aria-haspopup="true"
  onClick={[Function]}
  onContextMenu={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  Popup trigger
</button>
`;
